### PR TITLE
[BREAKING][Fix] Fix controller concurrency error when expanding production/consumption status

### PR DESCRIPTION
**Pull request by 0oshowero0**
## Background

The `TransferQueueController` uses two separate threads to handle metadata retrieval (`_process_request`) and production status updates (`_update_data_status`). This design is intended to enable real-time production status updates and prevent potential blocking of metadata retrieval (consumption) requests (since we do this in a zmq request loop).

```python3
    def _start_process_update_data_status(self):
        """Start the data status update processing thread."""
        self.process_update_data_status_thread = Thread(
            target=self._update_data_status,
            name="TransferQueueControllerProcessUpdateDataStatusThread",
            daemon=True,
        )
        self.process_update_data_status_thread.start()

    def _start_process_request(self):
        """Start the request processing thread."""
        self.process_request_thread = Thread(
            target=self._process_request, name="TransferQueueControllerProcessRequestThread", daemon=True
        )
        self.process_request_thread.start()
```

However, when combined with the dynamic expansion capability introduced in #98, the mask operation on production_status fails under concurrent execution:

```python3
        row_mask = torch.ones(self.total_samples_num, dtype=torch.bool)

        # Apply consumption filter (exclude already consumed samples)
        consumption_status = self.get_consumption_status(task_name)
        if consumption_status is not None:
            unconsumed_mask = consumption_status == 0
            row_mask &= unconsumed_mask

        # Create column mask for requested fields
        col_mask = torch.zeros(self.allocated_fields_num, dtype=torch.bool)
        field_indices = [self.field_name_mapping[field] for field in field_names]
        if field_indices:
            col_mask[field_indices] = True

        # !!! This will fail when produciton status expands
        relevant_status = self.production_status[row_mask][:, col_mask]      

        # Check if all required fields are ready for each sample
        all_fields_ready = torch.all(relevant_status, dim=1)
        ready_indices_in_filtered = torch.nonzero(all_fields_ready, as_tuple=False).flatten()

```

Concurrent execution of:
1. The status update thread (modifying production_status dimensions via dynamic expansion)
2. The metadata retrieval thread (performing mask indexing on production_status)

leads to dimension mismatch errors during the row_mask/col_mask indexing step.

## Solution

To resolve this race condition:

- Introduce a threading.Lock() for each individual DataPartitionStatus instance
- Acquire the lock when accessing/modifying consumption_status or production_status during:
    - Mask operation (indexing)
    - Dynamic dimension expansion

Since we do not need to strictly make sure the consumption request getting the newest status, we do not need to acquire the lock for the entire read/write operation, leading to better performance.

NOTE: Since we add a threading.Lock(), the original `get_partition` will fail when trying to get the `DataPartitionStatus` out. Therefore, we make the original method to be a inner function, and provide another `get_partition_snapshot` for users who needs to retrieve the consumption status for debug.

```python3
    # in class DataPartitionStatus
    def to_snapshot(self):
        """
        Get a snapshot of partition status information.

        Returns:
            DataPartitionStatus object without threading.Lock()
        """

        def _perform_copy():
            cls = self.__class__
            snapshot = cls.__new__(cls)

            for name, value in self.__dict__.items():
                if name == "data_status_lock":
                    continue

                if isinstance(value, torch.Tensor):
                    new_val = value.clone().detach()
                else:
                    new_val = copy.deepcopy(value)

                setattr(snapshot, name, new_val)
            return snapshot

        lock_obj = getattr(self, "data_status_lock", None)

        if lock_obj:
            with lock_obj:
                return _perform_copy()
        else:
            return _perform_copy()

```

```python3
    # in class TransferQueueController
    def _get_partition(self, partition_id: str) -> Optional[DataPartitionStatus]:
        """
        Get partition status information.

        Args:
            partition_id: ID of the partition to retrieve

        Returns:
            DataPartitionStatus object if partition exists, None otherwise
        """
        return self.partitions.get(partition_id)

    def get_partition_snapshot(self, partition_id: str) -> Optional[DataPartitionStatus]:
        """
        Get a copy of partition status information, without threading.Lock().

        Args:
            partition_id: ID of the partition to retrieve

        Returns:
            DataPartitionStatus object if partition exists, None otherwise
        """
        return self.partitions.get(partition_id).to_snapshot()
```

Thanks @NINGBENZHE for pointing out this issue!

CC: @LLLLxmmm 

## Summary by CodeRabbit

* **Refactor**
  * Improved synchronization mechanisms for data operations under concurrent access scenarios.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by 0oshowero0** [comment]
@coderabbitai review

File: transfer_queue/controller.py [hunk 1]
```diff
@@ -310,7 +314,8 @@ def update_production_status(
             required_samples = max_sample_idx + 1
 
             # Ensure we have enough rows
-            self.ensure_samples_capacity(required_samples)
+            with self.data_status_lock:
+                self.ensure_samples_capacity(required_samples)
```
[the remark this answers is not in the document]
Comment (0oshowero0) [reply transfer_queue/controller.py:332]: No need to strictly make sure this is protected
